base/tuple.jl

```diff
@@ -78,15 +78,20 @@ end
 eltype(t::Type{<:Tuple}) = _compute_eltype(t)
 function _compute_eltype(t::Type{<:Tuple})
     @_pure_meta
-    t isa Union && return typejoin(eltype(t.a), eltype(t.b))
+    t isa Union && return promote_join(eltype(t.a), eltype(t.b))
     t´ = unwrap_unionall(t)
     r = Union{}
     for ti in t´.parameters
-        r = typejoin(r, rewrap_unionall(unwrapva(ti), t))
+        r = promote_join(r, rewrap_unionall(unwrapva(ti), t))
     end
     return r
 end
 
+# promotion
+
+promote_rule(::Type{S}, ::Type{T}) where {S<:Tuple, T<:Tuple} =
```

Comment:
I think this should be removed. Changing the type of `[(1,), (1.0,)]` seems to be outside of the goal here.

Reply:
The goal is to have `[(1,), (missing,)] isa Vector{Tuple{Union{Int,Missing}}` rather than `Tuple{Any}`. I agree that's not totally required in this PR, but it makes a lot of sense to avoid an inconsistency with `map(identity, [(1,), (missing,)])`. That's also consistent with `[[1], [missing]]`. In general I don't see why tuples shoudn't implement promotion, like other containers.

Another solution is to use `promote_join` rather than `promote_type`. This could be done either here (i.e. special-case tuples promotion to just use `promote_join`), or more generally when concatenating arrays. I would prefer a solution which applies consistently to all types.

Anyway I've added a commit to revert this for now, so that we can address these issues separately.